DOCSP-42343: Secure Your Data + Authentication Mechanisms (#83)
mongodb/docs-scala: source/security/auth.txt

   :language: scala
   :copyable:
   :dedent:

Or, you can use a connection string that explicitly specifies
``authMechanism=SCRAM-SHA-1``:

.. literalinclude:: /includes/security/auth.scala
   :start-after: start-scram-sha-1-connection-string
   :end-before: end-scram-sha-1-connection-string
   :language: scala
   :copyable:
   :dedent:

X.509
-----

With the X.509 mechanism, MongoDB uses the X.509 certificate presented
during SSL negotiation to authenticate a user whose name is derived
from the distinguished name of the X.509 certificate.

X.509 authentication requires the use of SSL connections with
certificate validation. To create a credential of this type use the
``createMongoX509Credential()`` static factory method:

.. literalinclude:: /includes/security/auth.scala
   :start-after: start-mongodb-x509
   :end-before: end-mongodb-x509
   :language: scala
   :copyable:
   :dedent:

Or, you can use a connection string that explicitly specifies
``authMechanism=MONGODB-X509``:

.. literalinclude:: /includes/security/auth.scala
   :start-after: start-mongodb-x509-connection-string
   :end-before: end-mongodb-x509-connection-string
   :language: scala
   :copyable:
   :dedent:

> See the :manual:`Use x.509 Certificates to Authenticate Clients </tutorial/configure-x509-client-authentication/>`

Reviewer: I think this last sentence also needs to change, since you no longer need to determine the subject name from the certificate.

MongoCredential
---------------

An authentication credential is represented as an instance of the
``MongoCredential`` class. The ``MongoCredential`` class includes
static factory methods for each of the supported authentication
mechanisms.

Default Authentication Mechanism
--------------------------------

In MongoDB 3.0, MongoDB changed the default authentication mechanism
from ``MONGODB-CR`` to ``SCRAM-SHA-1``. In MongoDB 4.0, support for
the deprecated ``MONGODB-CR`` mechanism was removed and ``SCRAM-SHA-256`` support was
added.

To create a credential that authenticates by using the default
authentication mechanism, regardless of server version, create a
credential by using the ``createCredential()`` static factory method:

.. literalinclude:: /includes/security/auth.scala
   :start-after: start-default
   :end-before: end-default
   :language: scala
   :copyable:
   :dedent:

Or, you can use a connection string without explicitly specifying the
authentication mechanism:

.. literalinclude:: /includes/security/auth.scala
   :start-after: start-default-connection-string
   :end-before: end-default-connection-string
   :language: scala
   :copyable:
   :dedent:

> For challenge and response mechanisms, using the default
> authentication mechanism is the recommended approach, as it makes
> upgrading from MongoDB 2.6 to MongoDB 3.0 more simple, even after
> upgrading the authentication schema. For MongoDB 4.0 users, using the
> default authentication mechanism is also recommended as the mechanisms are
> checked and the correct hashing algorithm is used.

Reviewer 2: [s] Can potentially edit this to be more concise per [the style guide](https://www.mongodb.com/docs/meta/style-guide/terminology/concise-terms/) and use active voice. Something like: 
```suggestion
For challenge and response mechanisms, we recommend using the default
authentication mechanism.  This approach simplifies upgrading from MongoDB 
2.6 to MongoDB 3.0, even after upgrading the authentication schema. For 
MongoDB 4.0 users, the default authentication mechanism is also 
recommended because it checks the mechanisms and uses the correct hashing algorithm.
```

To explicitly create a credential of type ``SCRAM-SHA-256``, use
the ``createScramSha256Credential()`` method:

.. literalinclude:: /includes/security/auth.scala
   :start-after: start-scram-sha-256
   :end-before: end-scram-sha-256
   :language: scala
   :copyable:
   :dedent:

Or, you can use a connection string that explicitly specifies
``authMechanism=SCRAM-SHA-256``:

.. literalinclude:: /includes/security/auth.scala
   :start-after: start-scram-sha-256-connection-string
   :end-before: end-scram-sha-256-connection-string
   :language: scala
   :copyable:
   :dedent:

SCRAM-SHA-1
~~~~~~~~~~~

To explicitly create a credential of type ``SCRAM-SHA-1``, use the
``createScramSha1Credential()`` method:

.. literalinclude:: /includes/security/auth.scala
   :start-after: start-scram-sha-1
   :end-before: end-scram-sha-1
   :language: scala
   :copyable:
   :dedent:

Or, you can use a connection string that explicitly specifies
``authMechanism=SCRAM-SHA-1``:

.. literalinclude:: /includes/security/auth.scala
   :start-after: start-scram-sha-1-connection-string
   :end-before: end-scram-sha-1-connection-string
   :language: scala
   :copyable:
   :dedent:

X.509
-----

> With the ``X.509`` mechanism, MongoDB uses the X.509 certificate presented

Reviewer 2: [s] If the first mention uses monospace, it makes sense for the other mentions to as well for consistency for example: 
```suggestion
With the ``X.509`` mechanism, MongoDB uses the ``X.509`` certificate presented
```
Or to remove from the first mention.

Author: Will remove it from the first mention since it seems to be the standard in other documentation.
